File: src/2015/day-03.ts
import input from './files/input-day-03'
import { printResult } from '../shared/utils'

/**
 * Day 03
 * Instructions: https://adventofcode.com/2015/day/3
 */

type Direction = '^' | '>' | '<' | 'v'
type Coordinates = [number, number]

const getInputData = (): string => input.split('\n').filter((v) => v)[0]

function updateCoordinates(
  direction: Direction,
  coordinates: Coordinates,
): Coordinates {
  let [x, y] = coordinates

  if (direction === '^') {
    y += 1
  } else if (direction === '>') {
    x += 1
  } else if (direction === '<') {
    x -= 1
  } else if (direction === 'v') {
    y -= 1
  }

  return [x, y]
}

const solvePartOne = (directions: Direction[]): number => {
  const set = new Set().add('0,0')
  const coordinates: Coordinates = [0, 0]

  for (const direction of directions) {
    const [x, y] = updateCoordinates(direction, coordinates)
    coordinates[0] = x
    coordinates[1] = y
    set.add(coordinates.join(','))
  }

  return set.size
}

const solvePartTwo = (directions: Direction[]): number => {
  const santaCoordinates: Coordinates = [0, 0]
  const robotCoordinates: Coordinates = [0, 0]

  const coordinates = new Set()
  coordinates.add(santaCoordinates.join(','))
  coordinates.add(robotCoordinates.join(','))

  for (let i = 0; i < directions.length; i++) {
    const isSantaTurn = i % 2
    const direction = directions[i] as Direction

    if (isSantaTurn) {
      const [x, y] = updateCoordinates(direction, santaCoordinates)
      santaCoordinates[0] = x
      santaCoordinates[1] = y
      coordinates.add(santaCoordinates.join(','))
    } else {
      const [x, y] = updateCoordinates(direction, robotCoordinates)
      robotCoordinates[0] = x
      robotCoordinates[1] = y
      coordinates.add(robotCoordinates.join(','))
    }
  }

  return coordinates.size
}

export default () => {
  const inputData = getInputData().split('') as Direction[]

  const answerPartOne = solvePartOne(inputData)
  const answerPartTwo = solvePartTwo(inputData)

  printResult(answerPartOne, 2015, 3, 1)
  printResult(answerPartTwo, 2015, 3, 2)
}

File: csharp/2015/day-03.cs
internal enum Direction
{
    Up = '^',
    Right = '>',
    Left = '<',
    Down = 'v',
}

public class _2015_03
{
    private (int, int) UpdateCoordinates(char direction, (int x, int y) coordinates)
    {
        switch (direction)
        {
            case '^':
                return (coordinates.x, (coordinates.y += 1));
            case '<':
                return ((coordinates.x += 1), coordinates.y);
            case '>':
                return ((coordinates.x -= 1), coordinates.y);
            case 'v':
                return (coordinates.x, (coordinates.y -= 1));
            default:
                return (coordinates.x, coordinates.y);
        }
    }

    private int SolvePartOne(List<char> directions)
    {
        var set = new HashSet<string> { "0,0" };
        var coordinates = (0, 0);

        foreach (char direction in directions)
        {
            coordinates = UpdateCoordinates(direction, coordinates);
            set.Add($"{coordinates.Item1},{coordinates.Item2}");
        }

        return set.Count;
    }

    private int SolvePartTwo(List<char> directions)
    {
        var set = new HashSet<string> { "0,0", "0,0" };
        var map = new Dictionary<string, (int x, int y)>
        {
            { "santa", (0, 0) },
            { "robot", (0, 0) },
        };

        for (int i = 0; i < directions.Count; i++)
        {
            char direction = directions[i];
            string key = i % 2 == 0 ? "santa" : "robot";

            (int, int) coordinates = UpdateCoordinates(direction, map[key]);

            map[key] = coordinates;

            set.Add($"{map[key].Item1},{map[key].Item2}");
        }

        return set.Count;
    }

    public void run()
    {
        const string inputFileDate = "2015-03";

        var directions = FileUtils.GetInputData(inputFileDate)[0].ToList();

        if (directions.Count <= 0)
        {
            Console.WriteLine("No input data found for this day");
            return;
        }

        var answerPartOne = SolvePartOne(directions);
        var answerPartTwo = SolvePartTwo(directions);

        Utils.PrintResult(answerPartOne, 2015, 3, 1);
        Utils.PrintResult(answerPartTwo, 2015, 3, 2);
    }
}

